[Futures] set completed_ after postMarkCompletedHook

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54476 [Futures] Bump log verbosity when ignoring cb errors in python future.
* **#54475 [Futures] set completed_ after postMarkCompletedHook**

If completed_ is set before postMarkCompletedHook, errors in postMarkCompletedHook may not be correctly propagated as when we call `setError`, we throw an error claiming the future is already completed.

Since `postMarkCompletedHook` is additional processing installed by the framework for e.g. proper cuda stream handling, it makes sense that we only consider a future as completed_ after this processing has completed.

Differential Revision: [D27252965](https://our.internmc.facebook.com/intern/diff/D27252965/)

rohan-varma committed:
[Futures] enhance error handling in then()

Implements the proposal in https://github.com/pytorch/pytorch/issues/53717#issuecomment-800545655. See that issue for more details, but at a high level:

1. markCompleted() immediately sets completed_ = true
2. Subclasses of future (such as cuda future) implement a nontrivial `postMarkCompletedHook` which may throw
3. If above error is caught and we call `setError`, setError itself will error out because completed_ = true.

To fix this, only call setError if the user-defined cb resulted in an error, otherwise, call `markCompleted` and let postMarkCompletedHook() throw and crash the program (per @lw's thoughts this should be a fatal).

Differential Revision: [D27252965](https://our.internmc.facebook.com/intern/diff/D27252965/)

[ghstack-poisoned]

lw commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -468,11 +468,21 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
-            fut->markCompleted(cb());
+            cbValue= cb();
           } catch (std::exception&) {
+            hasException = true;
             fut->setError(std::current_exception());
```
Tip: by returning just after calling `setError` we could do without the boolean.

lw commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -468,11 +468,21 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
             fut->setError(std::current_exception());
           }
+          if (!hasException) {
+            fut->markCompleted(std::move(cbValue));
```
Yeah the reason I suggested catching exceptions here and fataling is that it's not fully clear to me what would happen to uncaught exceptions. I suspect they would "bubble up" to whoever called `markCompleted`, which might be a random thread in the thread pool, which may even be configured to catch, log and then ignore exceptions. I'm afraid that people don't watch logs unless something really goes wrong but, in this case, the "only" thing that would happen is an incorrect CUDA synchronization, which is a silent correctness issue. I'd rather we notify the user about such an issue as soon as we can, even if it means being very harsh.

rohan-varma replied:
Yeah, I think it would just show up in the logs. Regarding the thread pool, I believe that the c10 thread pool swallows and logs: https://github.com/pytorch/pytorch/blob/master/c10/core/thread_pool.cpp#L112, so it would only be surfacing to the user via logging as you mentioned.

rohan-varma commented:
cc @pritamdamania87 Does this one conflict with https://github.com/pytorch/pytorch/pull/54464 or should we land this one as well?

pritamdamania87 commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -468,11 +468,21 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
+            cbValue= cb();
           } catch (std::exception&) {
+            hasException = true;
             fut->setError(std::current_exception());
```
Shouldn't calling `setErrorIfNeeded` here be sufficient? It would log the appropriate exception and wouldn't result in the weird assertion error.

rohan-varma replied:
@lw's comment here https://github.com/pytorch/pytorch/issues/53717#issuecomment-800545655 says why that wouldn't work, in particular:

> If we just replaced setError with setErrorIfNeeded in then the end result would be that the future is still marked as successfully completed, whereas the CUDA part isn't set-up correctly, which could lead to silent CUDA synchronization/correctness issues.

lw replied:
Maybe let's add a comment in the code for why we can't use `setErrorIfNeeded`, to warn the next person that comes along and wonders the same thing?

pritamdamania87 replied:
I see, would a simpler fix be to set `completed_ = true` after `postMarkCompletedHook`? If we don't want to mark a future complete without running the `postMarkCompletedHook` successfully, then probably we should move `completed_ = true` accordingly?

rohan-varma commented:
Per offline discussion with @pritamdamania87 , switching this to just change where we set completed_ flag.

rohan-varma committed:
Update on "[Futures] set completed_ after postMarkCompletedHook"


If completed_ is set before postMarkCompletedHook, errors in postMarkCompletedHook may not be correctly propagated as when we call `setError`, we throw an error claiming the future is already completed.

Since `postMarkCompletedHook` is additional processing installed by the framework for e.g. proper cuda stream handling, it makes sense that we only consider a future as completed_ after this processing has completed.

Differential Revision: [D27252965](https://our.internmc.facebook.com/intern/diff/D27252965/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Futures] set completed_ after postMarkCompletedHook"


If completed_ is set before postMarkCompletedHook, errors in postMarkCompletedHook may not be correctly propagated as when we call `setError`, we throw an error claiming the future is already completed.

Since `postMarkCompletedHook` is additional processing installed by the framework for e.g. proper cuda stream handling, it makes sense that we only consider a future as completed_ after this processing has completed.

Differential Revision: [D27252965](https://our.internmc.facebook.com/intern/diff/D27252965/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Futures] set completed_ after postMarkCompletedHook"


If completed_ is set before postMarkCompletedHook, errors in postMarkCompletedHook may not be correctly propagated as when we call `setError`, we throw an error claiming the future is already completed.

Since `postMarkCompletedHook` is additional processing installed by the framework for e.g. proper cuda stream handling, it makes sense that we only consider a future as completed_ after this processing has completed.

Differential Revision: [D27252965](https://our.internmc.facebook.com/intern/diff/D27252965/)

[ghstack-poisoned]

lw reviewed (commented):
This could work too. Just two nits:
- perhaps we also want delay setting `value_` until after `postmarkCompletedHook` has run, so that we don't leave the future in a weird state where it's not complete and yet it holds a value (which could potentially keep some memory alive, making it look like it leaked)?
- at this point, should we rename `postMarkCompletedHook` to `preMarkCompletedHook`?

rohan-varma commented:
@lw Sure, both those suggestions sound reasonable. I'm assuming that means the new contract is that `postMarkCompletedHook` should run successfully, otherwise the future will be marked with an error?

rohan-varma committed:
Update on "[Futures] set completed_ after postMarkCompletedHook"


If completed_ is set before postMarkCompletedHook, errors in postMarkCompletedHook may not be correctly propagated as when we call `setError`, we throw an error claiming the future is already completed.

Since `postMarkCompletedHook` is additional processing installed by the framework for e.g. proper cuda stream handling, it makes sense that we only consider a future as completed_ after this processing has completed.

Differential Revision: [D27252965](https://our.internmc.facebook.com/intern/diff/D27252965/)

[ghstack-poisoned]

lw commented:
> I'm assuming that means the new contract is that postMarkCompletedHook should run successfully, otherwise the future will be marked with an error?

I thought the reason we were doing all of this was because we wanted to be able to tolerate the case of `postMarkCompletedHook` throwing an exception, isn't it?

As to what to do when that happens, I see two options:
- just let the exception bubble up, so that it reaches whoever was calling `markAsComplete` on the future, with the future still being in a "clean" (non-completed and non-error) state, and let them deal with it
- catch that exception and set the future to an error state

In practice there may not be much difference, since for example in the case of `then`, the `then` method itself will catch any unhandled exception and mark the future as errored.

rohan-varma commented:
> > I'm assuming that means the new contract is that postMarkCompletedHook should run successfully, otherwise the future will be marked with an error?
> 
> I thought the reason we were doing all of this was because we wanted to be able to tolerate the case of `postMarkCompletedHook` throwing an exception, isn't it?
> 
> As to what to do when that happens, I see two options:
> 
> * just let the exception bubble up, so that it reaches whoever was calling `markAsComplete` on the future, with the future still being in a "clean" (non-completed and non-error) state, and let them deal with it
> * catch that exception and set the future to an error state
> 
> In practice there may not be much difference, since for example in the case of `then`, the `then` method itself will catch any unhandled exception and mark the future as errored.

Yes, the main goal is to improve the handling of postMarkCompletedHook throwing an error. With this change, if it throws an error then `then()` should appropriately handle it as `setError` won't fail since the future is not marked as completed, and we'll set an error on the future. As for the options, it seems like it would be best to handle errors by setting them future and trying to avoid crashes?

rohan-varma committed:
Update on "[Futures] set completed_ after postMarkCompletedHook"


If completed_ is set before postMarkCompletedHook, errors in postMarkCompletedHook may not be correctly propagated as when we call `setError`, we throw an error claiming the future is already completed.

Since `postMarkCompletedHook` is additional processing installed by the framework for e.g. proper cuda stream handling, it makes sense that we only consider a future as completed_ after this processing has completed.

Differential Revision: [D27252965](https://our.internmc.facebook.com/intern/diff/D27252965/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Futures] set completed_ after postMarkCompletedHook"


If completed_ is set before postMarkCompletedHook, errors in postMarkCompletedHook may not be correctly propagated as when we call `setError`, we throw an error claiming the future is already completed.

Since `postMarkCompletedHook` is additional processing installed by the framework for e.g. proper cuda stream handling, it makes sense that we only consider a future as completed_ after this processing has completed.

Differential Revision: [D27252965](https://our.internmc.facebook.com/intern/diff/D27252965/)

[ghstack-poisoned]
